# Motivation

I was never trained to be a good programmer. I somehow passed the programming exam in my education and did not learn too much. I wanted to concque this challenage once for all to prove myself: no matter how low one's natual ability is, one can conquer anything with hardwork and persistent and it is never too late!

[Reference](https://www.techinterviewhandbook.org/grind75/faq)

# Week 1

##  (1) Two Sum

Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

 

Example 1:
```
Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
```
Example 2:

```
Input: nums = [3,2,4], target = 6
Output: [1,2]
```

Example 3:

```
Input: nums = [3,3], target = 6
Output: [0,1]
```
 
Constraints:

* 2 <= nums.length <= 104
* -109 <= nums[i] <= 109
* -109 <= target <= 109
* Only one valid answer exists.

Follow-up: Can you come up with an algorithm that is less than O(n^2) time complexity?

Using hashtable can get 
* time: $O(n)$
* space: $O(n)$

**key takeaway**

Use hashtable to store sum subtract current item to get desired item as we iterate the list to improve time complexity

In [2]:
from typing import List, Optional

class Solution:
    def twoSum1(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                num1 = nums[i]
                num2 = nums[j]
                if num2 == target - num1:
                    return [i, j]
    
    def twoSum2(self, nums, target):
        lookup = {}
        for i, num in enumerate(nums):
            if target - num in lookup:
                return [lookup[target - num], i]
            lookup[num] = i

s = Solution()

%timeit -n5 s.twoSum1([1, 3, 4, 8, 10, 2,7,11,15], 22)
%timeit -n5 s.twoSum2([1, 3, 4, 8, 10, 2,7,11,15], 22)

7.79 µs ± 347 ns per loop (mean ± std. dev. of 7 runs, 5 loops each)
1.72 µs ± 335 ns per loop (mean ± std. dev. of 7 runs, 5 loops each)


## (167) Two Sum II - Input Array is Sorted

Given a 1-indexed array of integers numbers that is already sorted in non-decreasing order, find two numbers such that they add up to a specific target number. Let these two numbers be `numbers[index1]` and `numbers[index2]` where `1 <= index1 < index2 < numbers.length`.

Return the indices of the two numbers, index1 and index2, added by one as an integer array [index1, index2] of length 2.

The tests are generated such that there is exactly one solution. You may not use the same element twice.

Your solution must use only constant extra space.

 

Example 1:
```
Input: numbers = [2,7,11,15], target = 9
Output: [1,2]
Explanation: The sum of 2 and 7 is 9. Therefore, index1 = 1, index2 = 2. We return [1, 2].
```
Example 2:
```
Input: numbers = [2,3,4], target = 6
Output: [1,3]
Explanation: The sum of 2 and 4 is 6. Therefore index1 = 1, index2 = 3. We return [1, 3].
```
Example 3:
```
Input: numbers = [-1,0], target = -1
Output: [1,2]
Explanation: The sum of -1 and 0 is -1. Therefore index1 = 1, index2 = 2. We return [1, 2].
``` 

Constraints:

* 2 <= numbers.length <= 3 * 104
* -1000 <= numbers[i] <= 1000
* numbers is sorted in non-decreasing order.
* -1000 <= target <= 1000
* The tests are generated such that there is exactly one solution.

Big O
* Time complexity $O(N)$
* Space complexity $O(1)$

**key takeaway**
* Using 2 pointer without hashtable, take advantage of sorted list
* User 2 pointers, one from the low end, one from the high end, if sum is < target, increase low index, if num is > target decrease high index

In [189]:
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        i = 0
        j = len(numbers) - 1
        #while i < len(numbers) and j >= 0:
        while i < j:
            if numbers[i] + numbers[j] < target:
                i+=1
            elif numbers[i] + numbers[j] > target:
                j-=1
            else:
                return [i+1, j+1]
            
numbers = [2,7,11,15]
target = 9
s = Solution()
s.twoSum(numbers, target)

[1, 2]

## (20) Valid Parentheses

Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Every close bracket has a corresponding open bracket of the same type.
 

Example 1:

```
Input: s = "()"
Output: true
```

Example 2:
```
Input: s = "()[]{}"
Output: true
```
Example 3:
```
Input: s = "(]"
Output: false
``` 

Constraints:

* 1 <= s.length <= 104
* s consists of parentheses only '()[]{}'.

Big O
* Time $O(n)$
* Space $O(n)$

**Key takeaaway**

Use push open parenthese to a stack, pop if matching, at the end if stack is empty, return true

In [8]:
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) <= 1:
            return False
        
        stack = []
        for c in s:
            if c == "(" or c == "[" or c == "{":
                stack.append(c)
            elif c == ")":
                if len(stack) == 0 or stack.pop(-1) != "(":
                    return False
            elif c == "}":
                if len(stack) == 0 or stack.pop(-1) != "{":
                    return False
            elif c == "]":
                if len(stack) == 0 or stack.pop(-1) != "[":
                    return False
        return len(stack) == 0

## (21) Merge Two Sorted Lists

You are given the heads of two sorted linked lists list1 and list2.

Merge the two lists in a one sorted list. The list should be made by splicing together the nodes of the first two lists.

Return the head of the merged linked list.

Example 1:

```
Input: list1 = [1,2,4], list2 = [1,3,4]
Output: [1,1,2,3,4,4]
```
Example 2:

```
Input: list1 = [], list2 = []
Output: []
```
Example 3:
```
Input: list1 = [], list2 = [0]
Output: [0]
```

Constraints:

* The number of nodes in both lists is in the range [0, 50].
* -100 <= Node.val <= 100
* Both list1 and list2 are sorted in non-decreasing order.

Iterative:
* time: $O(n+m)$
* space: $O(1)$

Recursive:
* time: $O(n+m)$
* space: $O(n+m)$

**key takeaway**
either use recursive as merge list with space cost or just insert each node from one list to the other

In [39]:
#Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists_recursive(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 is None:
            return list2
        if list2 is None:
            return list1
        
        if list1.val < list2.val:
            list1.next = self.mergeTwoList(list1.next, list2)
            return list1
        else:
            list2.next = self.mergeTwoLists(list1, list2.next)
            return list2
        
        
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        new_list = ListNode()
        head = new_list
        tail = new_list
        
        while list1 is not None and list2 is not None:
            if list1.val < list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            
            tail = tail.next
        
        if list1 is None:
            tail.next = list2
        else:
            tail.next = list1
            
        return head.next
            
list1 = ListNode(1, ListNode(2, ListNode(4))) 
list2 = ListNode(1, ListNode(3, ListNode(4))) 

s = Solution()
n = s.mergeTwoLists_recursive(list1, list2)
while n is not None:
    print(n.val)
    n = n.next

1
1
2
3
4
4


## (121) Best Time to buy and sell stock

You are given an array prices where prices[i] is the price of a given stock on the ith day.

You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

 

Example 1:
```
Input: prices = [7,1,5,3,6,4]
Output: 5
Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.
```
Example 2:
```
Input: prices = [7,6,4,3,1]
Output: 0
Explanation: In this case, no transactions are done and the max profit = 0.
```
 
Constraints:

* 1 <= prices.length <= 105
* 0 <= prices[i] <= 104

Big O
* Time $O(n)$
* Space $O(1)$

**key takeway**
Use a 2 pointer left and right (sliding window) update the pointer based on the local min in the window, update max profit based on the local max of the window

In [56]:
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) <= 1:
            return 0
        l, r = 0, 1
        maxprofit = 0
        while r < len(prices):
            if prices[r] > prices[l]:
                profit = prices[r] - prices[l]
                maxprofit = max(maxprofit, profit)
            else:
                l = r
            r+=1
        return maxprofit
    
s = Solution()
print(s.maxProfit([7,1,5,3,6,4]))
print(s.maxProfit([7,6,4,3,1]))
%timeit -n5 s.maxProfit([7,1,5,3,6,4])
%timeit -n5 s.maxProfit([7,6,4,3,1])

5
0
3.94 µs ± 332 ns per loop (mean ± std. dev. of 7 runs, 5 loops each)
1.97 µs ± 206 ns per loop (mean ± std. dev. of 7 runs, 5 loops each)


## (125) Valid Palindrome


A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.

Example 1:

```
Input: s = "A man, a plan, a canal: Panama"
Output: true
Explanation: "amanaplanacanalpanama" is a palindrome.
```

Example 2:

```
Input: s = "race a car"
Output: false
Explanation: "raceacar" is not a palindrome.
```
Example 3:

```
Input: s = " "
Output: true
Explanation: s is an empty string "" after removing non-alphanumeric characters.
Since an empty string reads the same forward and backward, it is a palindrome.
```
 
Constraints:

* 1 <= s.length <= 2 * 105
* s consists only of printable ASCII characters.

Big O
* Time $O(n)$
* Space $O(n)$ to store cleaned string
* Use 2 pointers to get $O(1)$ space

**key takeaway**

Using string slice:
* Python regrex `re.sub(r'[^0-9a-zA-Z]', '', s)` to remove nonalphanumeric
* String reverse using slice `reversed = s[::-1]`
* Palindrome means `s == s[::-1] => True`

User 2 pointers:
* Start left, right pointer at both end
* skip nonalphanumeric char
* convert to lower and compare if not equal return false
* This approach which would save space with $O(1)$

In [75]:
import re

class Solution:
    def isPalindrome(self, s: str) -> bool:
        cleaned_s = re.sub(r'[^0-9a-zA-Z]+', '', s.lower())
        if len(cleaned_s) <= 1:
            return True
        
        if len(cleaned_s)%2 == 0:
            return cleaned_s[:len(cleaned_s)//2] == cleaned_s[-len(cleaned_s)//2:][::-1]
        else:
            return cleaned_s[:len(cleaned_s)//2] == cleaned_s[-len(cleaned_s)//2+1:][::-1]

    def isPalindrome(self, s):
        s_clr = re.sub(r'[^a-zA-Z0-9]', '', s).lower()
        
        if s_clr == s_clr[::-1]:
            return True
        else:
            return False

    def isPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1
        while left < right:
            if not s[left].isalnum():
                left += 1
            elif not s[right].isalnum():
                right -= 1
            elif s[left].lower() != s[right].lower():
                return False
            else:
                left += 1
                right -= 1
        return True

    # if isalnum() is not allowed we need to create one
    def alphaNum(self, c):
        return (ord('A') <= ord(c) <= ord['Z'] or
                ord('a') <= ord(c) <= ord['z'] or
               ord('0') <= ord(c) <= ord['9'] or
        
s = Solution()
s.isPalindrome("A man, a plan, a canal: Panama")

True

## (226) Invert Binary Tree

Given the root of a binary tree, invert the tree, and return its root.

 

Example 1:

```
Input: root = [4,2,7,1,3,6,9]
Output: [4,7,2,9,6,3,1]
```

Example 2:

```
Input: root = [2,1,3]
Output: [2,3,1]
```
Example 3:

```
Input: root = []
Output: []
```
 

Constraints:

* The number of nodes in the tree is in the range [0, 100].
* -100 <= Node.val <= 100

Big O

* Time $O(h)$ height (or $O(log2(n)$
* Space $O(n)$

**Key Takeaway**
Use recursion to perform DFS (understanding of DFS)

In [76]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root is None:
            return None
        
        temp = root.left
        root.left = root.right
        root.right = temp
        self.invertTree(root.left)
        self.invertTree(root.right)
        return root
    
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        
        right = self.invertTree(root.right)
        left = self.invertTree(root.left)
        root.left = right
        root.right = left
        return root

    # iteratively (use Queue)
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        
        queue = collections.deque([root])
        while queue:
            current = queue.popleft()
            current.left, current.right = current.right, current.left
            
            if current.left:
                queue.append(current.left)
            
            if current.right:
                queue.append(current.right)
        
        return root

## (242) Valid Anagram

Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.
 

Example 1:

```
Input: s = "anagram", t = "nagaram"
Output: true
```
Example 2:

```
Input: s = "rat", t = "car"
Output: false
```

Constraints:

* 1 <= s.length, t.length <= 5 * 10^4
* s and t consist of lowercase English letters.

Big O:

* If use sort, time: $O(nlogn)$ with space either $O(1)$ or $O(n)$ depending on sort algorithm
* If use hashmap to store count and compare, time: $O(s+t)$, and space $O(s+t)$

**Key Takeaway**
* use hashmap to store count for each char and compare
* check length first to improve performance
* We can also use sort but different big O.

In [ ]:
class Solution:
    # Approach 1: sorted and compare directly
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        
        if sorted(s) == sorted(t):
            return True
        else:
            return False
    
    # Approach 2: sorted and compare in a loop 
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        sorted_s = sorted(s)
        sorted_t = sorted(t)
        for i in range(len(s)):
            if sorted_s[i] != sorted_t[i]:
                return False
        return True
    
    # Count frequency of chars and compare
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        s_dict = {}
        t_dict = {}
        for i in range(len(s)):
            s_dict[s[i]] = 1 + s_dict.get(s[i], 0)
            t_dict[t[i]] = 1 + t_dict.get(t[i], 0)
            
        for c in s_dict:
            if s_dict[c] != t_dict.get(c, 0):
                return False
        return True
    
    # Count frequency of chars and compare
    def isAnagram(self, s: str, t: str) -> bool:
        return Counter(s) == Counter(t)

## (704) Binary Search

Given an array of integers nums which is sorted in ascending order, and an integer target, write a function to search target in nums. If target exists, then return its index. Otherwise, return -1.

You must write an algorithm with O(log n) runtime complexity.

 

Example 1:
```
Input: nums = [-1,0,3,5,9,12], target = 9
Output: 4
Explanation: 9 exists in nums and its index is 4
```
Example 2:
```
Input: nums = [-1,0,3,5,9,12], target = 2
Output: -1
Explanation: 2 does not exist in nums so return -1
```

Constraints:

* 1 <= nums.length <= 104
* -104 < nums[i], target < 104
* All the integers in nums are unique.
* nums is sorted in ascending order.

Big O:

* Time $O(log n)$

In [ ]:
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        size = len(nums)
        low = 0
        high = size - 1
        
        while low <= high:
            mid = int((low + high)/2)
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                high = mid - 1
            else:
                low = mid + 1
        return -1

## (733) Flood Fill

An image is represented by an m x n integer grid image where image[i][j] represents the pixel value of the image.

You are also given three integers sr, sc, and color. You should perform a flood fill on the image starting from the pixel image[sr][sc].

To perform a flood fill, consider the starting pixel, plus any pixels connected 4-directionally to the starting pixel of the same color as the starting pixel, plus any pixels connected 4-directionally to those pixels (also with the same color), and so on. Replace the color of all of the aforementioned pixels with color.

Return the modified image after performing the flood fill.

 

Example 1:

```
Input: image = [[1,1,1],[1,1,0],[1,0,1]], sr = 1, sc = 1, color = 2
Output: [[2,2,2],[2,2,0],[2,0,1]]
Explanation: From the center of the image with position (sr, sc) = (1, 1) (i.e., the red pixel), all pixels connected by a path of the same color as the starting pixel (i.e., the blue pixels) are colored with the new color.
Note the bottom corner is not colored 2, because it is not 4-directionally connected to the starting pixel.
```
Example 2:
```
Input: image = [[0,0,0],[0,0,0]], sr = 0, sc = 0, color = 0
Output: [[0,0,0],[0,0,0]]
Explanation: The starting pixel is already colored 0, so no changes are made to the image.
```

Constraints:

* m == image.length
* n == image[i].length
* 1 <= m, n <= 50
* 0 <= image[i][j], color < 216
* 0 <= sr < m
* 0 <= sc < n

**key takeways**

* A typical DFS problem with recursion. A good way to learn DFS
* The base case is only fill when the color is equal to the current color of initial pixel

In [88]:
class Solution:
    def dfs(self, image, r, c, old_color, new_color):
        if image[r][c] == old_color:
            image[r][c] = new_color
            if r >= 1:
                self.dfs(image, r-1, c, old_color, new_color)
            if r < len(image) - 1:
                self.dfs(image, r+1, c, old_color, new_color)
            if c >= 1:
                self.dfs(image, r, c-1, old_color, new_color)
            if c < len(image[0])-1:
                self.dfs(image, r, c+1, old_color, new_color)
        else:
            return
                
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        old_color = image[sr][sc]
        if old_color == color:
            return image
        else:
            self.dfs(image, sr, sc, old_color, color)
        return image
        
s = Solution()
image = [[1,1,1],[1,1,0],[1,0,1]]
sr = 1
sc = 1
color = 2
print(s.floodFill(image, sr, sc, color))

[[2, 2, 2], [2, 2, 0], [2, 0, 1]]


## (235) Lowest Common Ancestor of a Binary Search Tree

Given a binary search tree (BST), find the lowest common ancestor (LCA) node of two given nodes in the BST.

According to the definition of LCA on Wikipedia: “The lowest common ancestor is defined between two nodes p and q as the lowest node in T that has both p and q as descendants (where we allow a node to be a descendant of itself).”

 

Example 1:

```
Input: root = [6,2,8,0,4,7,9,null,null,3,5], p = 2, q = 8
Output: 6
Explanation: The LCA of nodes 2 and 8 is 6.
```
Example 2:

```
Input: root = [6,2,8,0,4,7,9,null,null,3,5], p = 2, q = 4
Output: 2
Explanation: The LCA of nodes 2 and 4 is 2, since a node can be a descendant of itself according to the LCA definition.
```
Example 3:

```
Input: root = [2,1], p = 2, q = 1
Output: 2
```
 
Constraints:

* The number of nodes in the tree is in the range [2, 105].
* -109 <= Node.val <= 109
* All Node.val are unique.
* p != q
* p and q will exist in the BST.

Big O

* if use recursion, time is $O(N)$, space is $O(N)$
* if do not use recursion, time is $O(logn)$, space is $O(1)$

**Key Takeaway**
* binary tree property based on the comparing p, q value to the current root value
* The base case is current root value

In [91]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # Interative
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        cur = root
        while cur is not None:
            if p.val > cur.val and q.val > cur.val:
                cur = cur.right
            elif p.val < cur.val and q.val < cur.val:
                cur = cur.left
            else:
                return cur

    # recursive
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # Value of current node or parent node.
        parent_val = root.val

        # Value of p
        p_val = p.val

        # Value of q
        q_val = q.val

        # If both p and q are greater than parent
        if p_val > parent_val and q_val > parent_val:    
            return self.lowestCommonAncestor(root.right, p, q)
        # If both p and q are lesser than parent
        elif p_val < parent_val and q_val < parent_val:    
            return self.lowestCommonAncestor(root.left, p, q)
        # We have found the split point, i.e. the LCA node.
        else:
            return root

## (110) Balanced Binary Tree

Given a binary tree, determine if it is height-balanced

Example 1:

```
Input: root = [3,9,20,null,null,15,7]
Output: true
```
Example 2:

```
Input: root = [1,2,2,3,3,null,null,4,4]
Output: false
```
Example 3:
```
Input: root = []
Output: true
```

Constraints:

* The number of nodes in the tree is in the range [0, 5000].
* -104 <= Node.val <= 104

In [ ]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def height(self, root):
        if root is None:
            return -1
        return 1 + max(height(root.left), height(root.right))
        
        
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        if abs(self.height(root.left) - self.height(root.right))>1:
            return False
        else:
            return self.isBalanced(root.left) && self.isBalanced(root.right)

class Solution:
    # Return whether or not the tree at root is balanced while also returning
    # the tree's height
    def isBalancedHelper(self, root: TreeNode) -> (bool, int):
        # An empty tree is balanced and has height -1
        if not root:
            return True, -1
        
        # Check subtrees to see if they are balanced. 
        leftIsBalanced, leftHeight = self.isBalancedHelper(root.left)
        if not leftIsBalanced:
            return False, 0
        rightIsBalanced, rightHeight = self.isBalancedHelper(root.right)
        if not rightIsBalanced:
            return False, 0
        
        # If the subtrees are balanced, check if the current tree is balanced
        # using their height
        return (abs(leftHeight - rightHeight) < 2), 1 + max(leftHeight, rightHeight)
        
    def isBalanced(self, root: TreeNode) -> bool:
        return self.isBalancedHelper(root)[0]

## (141) Linked List Cycle

Given head, the head of a linked list, determine if the linked list has a cycle in it.

There is a cycle in a linked list if there is some node in the list that can be reached again by continuously following the next pointer. Internally, pos is used to denote the index of the node that tail's next pointer is connected to. Note that pos is not passed as a parameter.

Return true if there is a cycle in the linked list. Otherwise, return false.

 

Example 1:

```
Input: head = [3,2,0,-4], pos = 1
Output: true
Explanation: There is a cycle in the linked list, where the tail connects to the 1st node (0-indexed).
```
Example 2:

```
Input: head = [1,2], pos = 0
Output: true
Explanation: There is a cycle in the linked list, where the tail connects to the 0th node.
```
Example 3:

```
Input: head = [1], pos = -1
Output: false
Explanation: There is no cycle in the linked list.
```

Constraints:

* The number of the nodes in the list is in the range [0, 104].
* -105 <= Node.val <= 105
* pos is -1 or a valid index in the linked-list.


Big O
* Time: $O(n)$
* Space $O(1)$ if we use 2 pointer (fast, slow) $O(n)$ if we use hashmap

**Key takeaway**
* use a fast and slow pointer, fast is 2x faster than slow, this utilize Floyd's cycle-finding algorithm, also known as the "tortoise and hare" algorithm, to determine the presence of a loop efficiently.
* If there is a loop in the list, the fast pointer will eventually meet the slow pointer.
* Or use a hashmap if it is there in the hashmap, we find the cycle

In [ ]:
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False
        
        cur = head
        hashmap = {}
        while cur is not None:
            hashmap[cur] = 1+hashmap.get(cur, 0)
            cur = cur.next
            if hashmap.get(cur, 0) > 1:
                return True
        return False
    
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False
        
        slow, fast = head, head
        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                return True
        return False

## (232) Implement Queue Using Stacks

Implement a first in first out (FIFO) queue using only two stacks. The implemented queue should support all the functions of a normal queue (push, peek, pop, and empty).

Implement the MyQueue class:

* `void push(int x)` Pushes element x to the back of the queue.
* `int pop()` Removes the element from the front of the queue and returns it.
* `int peek()` Returns the element at the front of the queue.
* `boolean empty()` Returns true if the queue is empty, false otherwise.

Notes:

* You must use only standard operations of a stack, which means only push to top, peek/pop from top, size, and is empty operations are valid.
* Depending on your language, the stack may not be supported natively. You may simulate a stack using a list or deque (double-ended queue) as long as you use only a stack's standard operations.
 

Example 1:

```
Input
["MyQueue", "push", "push", "peek", "pop", "empty"]
[[], [1], [2], [], [], []]
Output
[null, null, null, 1, 1, false]
```

Explanation
```
MyQueue myQueue = new MyQueue();
myQueue.push(1); // queue is: [1]
myQueue.push(2); // queue is: [1, 2] (leftmost is front of the queue)
myQueue.peek(); // return 1
myQueue.pop(); // return 1, queue is [2]
myQueue.empty(); // return false
```
 

Constraints:

* 1 <= x <= 9
* At most 100 calls will be made to push, pop, peek, and empty.
* All the calls to pop and peek are valid.

Follow-up: Can you implement the queue such that each operation is amortized O(1) time complexity? In other words, performing n operations will take overall O(n) time even if one of those operations may take longer.

Big O
* Time: $O(n)$ for push or for pop based on approach

**key takeaway**

use the second stack to order or the queue when adding a new element

In [117]:
class MyQueue: # push O(n), pop O(1)

    def __init__(self):
        self.stack1 = []
        self.stack2 = []

    def push(self, x: int) -> None:
        while len(self.stack1) > 0:
            self.stack2.append(self.stack1.pop(-1))
         
        self.stack1.append(x)
        
        while len(self.stack2) > 0:
            self.stack1.append(self.stack2.pop(-1))
        

    def pop(self) -> int:
        if len(self.stack1) == 0:
            return None
        return self.stack1.pop(-1)

    def peek(self) -> int:
        if len(self.stack1) == 0:
            return None
        return self.stack1[-1]

    def empty(self) -> bool:
        return len(self.stack1) == 0

class MyQueue: # push O(1), pop O(n)
    def __init__(self):
        self.s1 = []
        self.s2 = []

    def push(self, x):
        self.s1.append(x)

    def pop(self):
        self.peek()
        return self.s2.pop()

    def peek(self):
        if not self.s2:
            while self.s1:
                self.s2.append(self.s1.pop())
        return self.s2[-1]        

    def empty(self):
        return not self.s1 and not self.s2

# Week 2

## (278) First Bad Version

You are a product manager and currently leading a team to develop a new product. Unfortunately, the latest version of your product fails the quality check. Since each version is developed based on the previous version, all the versions after a bad version are also bad.

Suppose you have n versions [1, 2, ..., n] and you want to find out the first bad one, which causes all the following ones to be bad.

You are given an API bool isBadVersion(version) which returns whether version is bad. Implement a function to find the first bad version. You should minimize the number of calls to the API.

 

Example 1:
```
Input: n = 5, bad = 4
Output: 4
Explanation:
call isBadVersion(3) -> false
call isBadVersion(5) -> true
call isBadVersion(4) -> true
Then 4 is the first bad version.
```
Example 2:
```
Input: n = 1, bad = 1
Output: 1
```

Constraints:

1 <= bad <= n <= 2^31 - 1

Big O:
time: $O(logn)$
space: $O(1)$

**key takeaway**

Use binary search to find with minimum call, do not add 1 to right (high) when find it a bad version because it needs to be in the next search space in case it is the last one.

In [ ]:
# The isBadVersion API is already defined for you.
def isBadVersion(version: int) -> bool:
    pass

class Solution:
    def firstBadVersion(self, n: int) -> int:
        if n == 1:
            return 1 if isBadVersion(1) else None
        low = 1
        high = n
        while low < high:
            mid = int((high + low)/2)
            if isBadVersion(mid):
                high = mid
            else:
                low = mid + 1
        return low

## (383) Ransom Note

Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.

Each letter in magazine can only be used once in ransomNote.


Example 1:
```
Input: ransomNote = "a", magazine = "b"
Output: false
```
Example 2:
```
Input: ransomNote = "aa", magazine = "ab"
Output: false
```
Example 3:
```
Input: ransomNote = "aa", magazine = "aab"
Output: true
```

Constraints:

* 1 <= ransomNote.length, magazine.length <= 105
* ransomNote and magazine consist of lowercase English letters.

Big O
* Time: $O(m)$
* Space $O(k)$ where k is the 26 (alphabets)

**key Takeaway**
Using hashmap to store available characters and counts, iterate ransomNote to decrease count if it depletes the characters store, it is false, otherwise true

In [3]:
import collections

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if len(magazine) < len(ransomNote):
            return False
        
        
        hashmap = {}
        for c in magazine:
            hashmap[c] = 1 + hashmap.get(c, 0)
        
        for c in ransomNote:
            if c not in hashmap:
                return False
            else:
                hashmap[c] -= 1
                if hashmap[c] < 0:
                    return False
        return True

    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if len(magazine) < len(ransomNote):
            return False
        
        
        letters = collections.Counter(magazine)
        for c in ransomNote:
            # If there are none of c left, return False.
            if letters[c] <= 0:
                return False
            # Remove one of c from the Counter.
            letters[c] -= 1
        return True
    
ransomNote = "chejaccdae"
magazine = "geceeibccchjejhdd"
s = Solution()
s.canConstruct(ransomNote, magazine)

False

## (70) Climbing Stairs

You are climbing a staircase. It takes n steps to reach the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

 
Example 1:
```
Input: n = 2
Output: 2
Explanation: There are two ways to climb to the top.
1. 1 step + 1 step
2. 2 steps
```
Example 2:
```
Input: n = 3
Output: 3
Explanation: There are three ways to climb to the top.
1. 1 step + 1 step + 1 step
2. 1 step + 2 steps
3. 2 steps + 1 step
``` 

Constraints:

* 1 <= n <= 45

Big O

* Time: $O(2^n)$ if use recursion too slow. $O(n) if we use dynamic programming


**key takeaway**
Use bottom-up dynamic programming

In [2]:
class Solution:
    def __init__(self):
        self.cache = {1:1, 2:2}

    # pure recursive
    def climbStairs_recursion(self, n: int) -> int:
        if n == 1:
            return 1
        elif n == 2:
            return 2
        else:
            return self.climbStairs_recursion(n-1) + self.climbStairs_recursion(n-2)
    
    # recursion with hashtable
    def climbStairs_cache(self, n: int) -> int:
        if n == 1:
            return 1
        if n == 2:
            return 2
        
        if n in self.cache:
            return self.cache[n]
        else:
            ways = self.climbStairs_cache(n-1) + self.climbStairs_cache(n-2)
            self.cache[n] = ways
            return ways
        
    def climbStairs_dp(self, n: int) -> int:
        if n == 1:
            return 1
        if n == 2:
            return 2
        
        N_2, N_1 = 1, 2
        for N in range(3, n+1):
            temp = N_1
            Next = N_2 + N_1
            N_1 = Next
            N_2 = temp
        return Next

s = Solution()
#%timeit -n5 s.climbStairs_recursion(30)
%timeit -n10 s.climbStairs_cache(100)
%timeit -n10 s.climbStairs_dp(100)

The slowest run took 26.96 times longer than the fastest. This could mean that an intermediate result is being cached.
2.23 µs ± 4.25 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
6.86 µs ± 1.18 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


## (409) Longest Palindrome

Given a string s which consists of lowercase or uppercase letters, return the length of the longest palindrome that can be built with those letters.

Letters are case sensitive, for example, "Aa" is not considered a palindrome here.

 

Example 1:
```
Input: s = "abccccdd"
Output: 7
Explanation: One longest palindrome that can be built is "dccaccd", whose length is 7.
```
Example 2:
```
Input: s = "a"
Output: 1
Explanation: The longest palindrome that can be built is "a", whose length is 1.
```

Constraints:

* 1 <= s.length <= 2000
* s consists of lowercase and/or uppercase English letters only.

Big O

* Time O(n)
* Space O(1)

**Key takeaway**

Use a hashtable to store character frequency, determine whether there is a unique center or not if yes need to add 1

In [9]:
class Solution:
    def longestPalindrome(self, s: str) -> int:
        store = collections.Counter(s)
        even_size = 0
        odd_size = 0
        odd_count = 0
        for c, v in store.items():
            if v%2 == 0:
                even_size += v
            else:
                odd_size += v - 1
                odd_count+=1
        if odd_count == 0:
            return even_size
        else:
            return even_size + odd_size + 1        

## (206) Reverse Linked List

Given the head of a singly linked list, reverse the list, and return the reversed list.

 

Example 1:

```
Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]
```
Example 2:

```
Input: head = [1,2]
Output: [2,1]
```
Example 3:
```
Input: head = []
Output: []
``` 

Constraints:

* The number of nodes in the list is the range [0, 5000].
* -5000 <= Node.val <= 5000

Big O
* time $O(n)$
* mem $O(1)$

**Key takeaway**

create current (set to head) and prev (set to Null), 2 pointers, reverse the link and shift 2 pointer forward until the current read the end (become none), return prev since it is points to the last node

In [ ]:
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr = head
        prev = None
        while curr is not None:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        return prev

## (169) Majority Element

Given an array nums of size n, return the majority element.

The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array.

 

Example 1:
```
Input: nums = [3,2,3]
Output: 3
```
Example 2:
```
Input: nums = [2,2,1,1,1,2,2]
Output: 2
```

Constraints:

* n == nums.length
* 1 <= n <= 5 * 104
* -109 <= nums[i] <= 109

Big O
if use hashmap

* Time $O(n)$
* Space $O(n)$

if use sort

* Time $O(nlogn)$
* Space $O(1)$

**Key Takeaway**

* hashmap is straightfoward but more costly
* sort method: sort the list first, the majority element will always occupy the middle 1 or 2 index depending on the length
* **Boyer-Moore Voting Algorithm** is a the best approach
    * Use the first element as the candidate, keep a counter
    * iterate throw the list, if the next element is the same element +1, if different -1
    * when the counter = 0, the next element becomes the new candidate
    * In the end of iteration, the candidate will be the majority element

In [22]:
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        freq_counter = collections.Counter(nums)
        for k, v in freq_counter.items():
            if v == max(freq_counter.values()):
                return k
            
    def majorityElement(self, nums: List[int]) -> int:
        count = {}
        res, maxCount = 0, 0
        for n in nums:
            count[n] = 1 + count.get(n, 0)
            res = n if count[n] > maxCount else res
            maxCount = max(count[n], maxCount)
        return res
    
    def majorityElement(self, nums: List[int]) -> int:
        freq_counter = collections.Counter(nums)
        return max(freq_counter.keys(), key=counts.get)
    
    def majorityElement(self, nums: List[int]) -> int:
        majority_count = len(nums) // 2 + 1
        freq_counter = collections.Counter(nums)
        for k, v in freq_counter.items():
            if v >= majority_count:
                return k
    
    def majorityElement(self, nums):
        nums.sort()
        return nums[len(nums)//2]

    # Boyer-moore algorithm
    def majorityElement(self, nums: List[int]) -> int:
        count, candidate = 0, 0
        for n in nums:
            if count == 0:
                candidate = n
            count += (1 if n == candidate else - 1)
        return candidate
    

nums = [2, 2, 1, 1, 1, 2, 2]
s = Solution()
s.majorityElement(nums)

2

## (67) Add Binary

Given two binary strings a and b, return their sum as a binary string.


Example 1:
```
Input: a = "11", b = "1"
Output: "100"
```
Example 2:
```
Input: a = "1010", b = "1011"
Output: "10101"
```
 
Constraints:

* 1 <= a.length, b.length <= 104
* a and b consist only of '0' or '1' characters.
* Each string does not contain leading zeros except for the zero itself.

Big O

* Time $O(n)$
* Space $O(1)$

**key takeaway**

Similar to other add problem, keep a carry bits, reverse order or iterate reversely, use `ord()` function to calculate number using ascii values, total%2 naturally gives the char, total//2 naturally gives the carry

In [39]:
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        result = ''
        carry = 0
        max_len = max(len(a), len(b))
        a = a.zfill(max_len)
        b = b.zfill(max_len)
        
        for i in range(max_len - 1, -1 ,-1):
            num1 = 1 if a[i] == '1' else 0
            num2 = 1 if b[i] == '1' else 0
            result = ('1' if (num1 + num2 + carry) % 2 == 1 else '0') + result
            carry = 0 if num1 + num2 + carry < 2 else 1
        if carry != 0:
            result = '1' + result
        return result
    
    def addBinary(self, a: str, b: str) -> str:
        res = ''
        carry = 0
        a, b = a[::-1], b[::-1]
        for i in range(max(len(a), len(b))):
            digit_a = ord(a[i]) - ord("0") if i < len(a) else 0
            digit_b = ord(b[i]) - ord("0") if i < len(b) else 0
            
            total = digit_a + digit_b + carry
            char = total%2
            res = char + res
            carry = total//2
        if carry != 0:
            res = '1' + res
        return res
            
    # using bit manipulation
    # XOR is sum without carray
    # AND and shift 1 bits to the left find the carray
    # we keep shift carry to the left unit it becomes 0
    def addBinary(self, a, b) -> str:
        x, y = int(a, 2), int(b, 2)
        while y:
            answer = x ^ y
            carry = (x & y) << 1
            x, y = answer, carry
        return bin(x)[2:]
    

s = Solution()
s.addBinary(a='11', b = '1')
s.addBinary(a='1010', b = '101100')

'110110'

## (543) Diameter of Binary Tree

Given the root of a binary tree, return the length of the diameter of the tree.

The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or may not pass through the root.

The length of a path between two nodes is represented by the number of edges between them.

 

Example 1:

```
Input: root = [1,2,3,4,5]
Output: 3
Explanation: 3 is the length of the path [4,2,1,3] or [5,2,1,3].
```
Example 2:
```
Input: root = [1,2]
Output: 1
```
 

Constraints:

* The number of nodes in the tree is in the range [1, 104].
* -100 <= Node.val <= 100

Big O

* Time: $O(n)$
* Space $O(n)$

**key takeway**
* DFS
* initalize an integer variable diameter to keep track of the longest path we find from the DFS.
* Implement a recursive function longestPath which takes a TreeNode as input. It should recursively explore the entire tree rooted at the given node. Once it's finished, it should return the longest path out of its left and right branches:
* if node is None, we have reached the end of the tree, hence we should return 0;
* we want to recursively explore node's children, so we call longestPath again with node's left and right children. In return, we get the longest path of its left and right children leftPath and rightPath;
* if leftPath plus rightPath is longer than the current longest diameter found, then we need to update diameter;
* finally, we return the longer one of leftPath and rightPath. Remember to add 111 as the edge connecting it with its parent.
* Call longestPath with root.

In [ ]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = 0
        
        def traverse(node):
            if not node:
                return 0
            nonlocal diameter
            left = traverse(node.left)
            right = traverse(node.right)
           
            diameter = max(diameter, left + right)
            return max(left, right) + 1
        
        traverse(root)
        return diameter
            


s = Solution()
s.diameterOfBinaryTree

## (876) Middle of the Linked List

Given the head of a singly linked list, return the middle node of the linked list.

If there are two middle nodes, return the second middle node.

 

Example 1:

```
Input: head = [1,2,3,4,5]
Output: [3,4,5]
Explanation: The middle node of the list is node 3.
```
Example 2:

```
Input: head = [1,2,3,4,5,6]
Output: [4,5,6]
Explanation: Since the list has two middle nodes with values 3 and 4, we return the second one.
```

Constraints:

* The number of nodes in the list is in the range [1, 100].
* 1 <= Node.val <= 100

Big O

* Time: $O(n)$
* Space: $O(1)$

**key takeaway**

Use 2 pointer, slow and fast, fast go 2x as the slow, when fast reach the end, slow reach the middle

In [ ]:
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = head
        fast = head
        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow = slow.next
        
        return flow

## (104) Maximum Depth of Binary Tree

Given the root of a binary tree, return its maximum depth.

A binary tree's maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.

 

Example 1:

```
Input: root = [3,9,20,null,null,15,7]
Output: 3
```
Example 2:
```
Input: root = [1,null,2]
Output: 2
```
 

Constraints:

* The number of nodes in the tree is in the range [0, 104].
* -100 <= Node.val <= 100

Big O

* Time $O(logn)$
* space $O(logn)$

In [42]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        left_size, right_size = 0, 0
        if root.left is not None:
            left_size = self.maxDepth(node.left)
        if node.right is not None:
            right_size = self.maxDepth(node.right)
        return 1 + max(left_size, right_size)

## (217) Contains Duplicate

Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

 

Example 1:
```
Input: nums = [1,2,3,1]
Output: true
```
Example 2:
```
Input: nums = [1,2,3,4]
Output: false
```
Example 3:
```
Input: nums = [1,1,1,3,3,4,3,2,4,2]
Output: true
```
 

Constraints:

* 1 <= nums.length <= 105
* -109 <= nums[i] <= 109

Big O

* Time $O(nlogn)$ (use sorting)
* Space $O(1)$

**key takeaway**

We can use set but it may not be optimal. We can either use sorted first then linear search or use hashtable

In [44]:
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        return len(set(nums)) != len(nums)
    
    def containsDuplicate(self, nums: List[int]) -> bool:
        nums = sorted(nums)
        for i in range(1, len(nums)):
            if nums[i-1] == nums[i]:
                return True
        return False

## (53) Maximum Subarray -- (Kadane's Algorithm)

Given an integer array nums, find the subarray with the largest sum, and return its sum.

 

Example 1:
```
Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
Output: 6
Explanation: The subarray [4,-1,2,1] has the largest sum 6.
```
Example 2:
```
Input: nums = [1]
Output: 1
Explanation: The subarray [1] has the largest sum 1.
```
Example 3:
```
Input: nums = [5,4,-1,7,8]
Output: 23
Explanation: The subarray [5,4,-1,7,8] has the largest sum 23.
```
 
Constraints:

* 1 <= nums.length <= 10^5
* -10^4 <= nums[i] <= 10^4
 
Big O

* Time $O(n)$
* Space $O(1)$

**Key Takeaway**

* If we use brunt force, we need to iterate every combination and perform sum calculate => $O(n^3)$
* We can use cumulative sum to bring the complexity to $O(n^2)$ but still very slow
* Use dynamic programming to reduce $O(n)$
    * Current max is max(previous max + current number, current number)

**Whenever you see a question that asks for the maximum or minimum of something, consider Dynamic Programming as a possibility**

In [108]:
class Solution:
    # Brute Force with cumSum: Time limit exceed with O(n^2)
    def maxSubArray(self, nums: List[int]) -> int:
        max_val = nums[0]
        for i in range(len(nums)):
            curSum = 0
            for j in range(i, len(nums)):
                curSum += nums[j]
                max_val = max(max_val, curSum)
        return max_val
    
    # Dynamic programming
    def maxSubArray(self, nums: List[int]) -> int:
        max_val = [nums[0]]
        for i in range(1, len(nums)):
            max_val.append(max(max_val[i-1] + nums[i], nums[i]))
            #max_val.append(max(max_val[i-1], 0) + nums[i])
    
        return max(max_val)

    def maxSubArray(self, nums: List[int]) -> int:
        max_val = nums[0]
        curSum = nums[0]
        for i in range(1, len(nums)):
            curSum = max(curSum + nums[i], nums[i])
            #curSum = max(curSum, 0) + nums[i]
            max_val = max(max_val, curSum)
    
        return max_val
            
                      
s = Solution()
s.maxSubArray([5, 4, -1, 7, 8])
#s.maxSubArray([-2,1,-3,4,-1,2,1,-5,4])
#s.maxSubArray([-2,-3,-1])
#s.maxSubArray([-1,0,-2])

23

# Week 3

## (57) Insert Interval

You are given an array of non-overlapping intervals intervals where intervals[i] = [starti, endi] represent the start and the end of the ith interval and intervals is sorted in ascending order by starti. You are also given an interval newInterval = [start, end] that represents the start and end of another interval.

Insert newInterval into intervals such that intervals is still sorted in ascending order by starti and intervals still does not have any overlapping intervals (merge overlapping intervals if necessary).

Return intervals after the insertion.

 

Example 1:
```
Input: intervals = [[1,3],[6,9]], newInterval = [2,5]
Output: [[1,5],[6,9]]
```
Example 2:
```
Input: intervals = [[1,2],[3,5],[6,7],[8,10],[12,16]], newInterval = [4,8]
Output: [[1,2],[3,10],[12,16]]
Explanation: Because the new interval [4,8] overlaps with [3,5],[6,7],[8,10].
``` 

Constraints:

* 0 <= intervals.length <= 104
* intervals[i].length == 2
* 0 <= starti <= endi <= 105
* intervals is sorted by starti in ascending order.
* newInterval.length == 2
* 0 <= start <= end <= 105

In [118]:
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []
        for i in range(len(intervals)):
            # edge case 1: new interval is not overlapping with the current interval
            # and it is smaller then the first element in the interval
            if newInterval[1] < intervals[i][0]:
                res.append(newInterval)
                return res + intervals[i:]
            # new interval is not overlapping with current interval
            # it may overlapping with next one, we append the current interval
            elif newInterval[0] > intervals[i][1]:
                res.append(intervals[i])
            # there is overlapping
            else:
                newInterval = [min(newInterval[0], intervals[i][0]), max(newInterval[1], intervals[i][1])]
        res.append(newInterval) 
        return res
            

s = Solution()
s.insert([[1,3],[6,9]], [2, 5])
#s.insert([[1,2],[3,5],[6,7],[8,10],[12,16]], [4, 8])

[[1, 5], [6, 9]]

## (542) 01 Matrix

Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.

The distance between two adjacent cells is 1.

 

Example 1:

```
Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
Output: [[0,0,0],[0,1,0],[0,0,0]]
```
Example 2:

```
Input: mat = [[0,0,0],[0,1,0],[1,1,1]]
Output: [[0,0,0],[0,1,0],[1,2,1]]
```

Constraints:

* m == mat.length
* n == mat[i].length
* 1 <= m, n <= 104
* 1 <= m * n <= 104
* mat[i][j] is either 0 or 1.
* There is at least one 0 in mat.


**key takeaway**

Finding the nearst ... is usually use BFS, to improve efficiency, it is not aways search from the source to target but also search from the target to the source

**Create a BFS Template to use**

In [236]:
from collections import deque

class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        def valid(row, col):
            return 0 <= row < m and 0 <= col < n
        
        matrix = [row[:] for row in mat]
        m = len(matrix)
        n = len(matrix[0])
        queue = deque()
        seen = set()
        
        for row in range(m):
            for col in range(n):
                if matrix[row][col] == 0:
                    queue.append((row, col, 0))
                    seen.add((row, col))
        
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

        while queue:
            row, col, steps = queue.popleft()
            
            for dx, dy in directions:
                next_row, next_col = row + dy, col + dx
                if (next_row, next_col) not in seen and valid(next_row, next_col):
                    seen.add((next_row, next_col))
                    queue.append((next_row, next_col, steps + 1))
                    matrix[next_row][next_col] = steps + 1
        return matrix

## K Closest Points to Origin

Given an array of points where points[i] = [xi, yi] represents a point on the X-Y plane and an integer k, return the k closest points to the origin (0, 0).

The distance between two points on the X-Y plane is the Euclidean distance (i.e., √(x1 - x2)2 + (y1 - y2)2).

You may return the answer in any order. The answer is guaranteed to be unique (except for the order that it is in).

```
Input: points = [[1,3],[-2,2]], k = 1
Output: [[-2,2]]
Explanation:
The distance between (1, 3) and the origin is sqrt(10).
The distance between (-2, 2) and the origin is sqrt(8).
Since sqrt(8) < sqrt(10), (-2, 2) is closer to the origin.
We only want the closest k = 1 points from the origin, so the answer is just [[-2,2]].
```

Example 2:
```
Input: points = [[3,3],[5,-1],[-2,4]], k = 2
Output: [[3,3],[-2,4]]
Explanation: The answer [[-2,4],[3,3]] would also be accepted.
```

Constraints:

* 1 <= k <= points.length <= 104
* -104 < xi, yi < 104

Big O:
* Time $O(n)$ if use heap
* Time $O(nlogn)$ if use sort approach
* Space $O(n)$

**Key takeaway**

* Use min heap (python provided data structure) to store [distance, x, y] of each points, the pop k ties
* Just sort the points direction using a lambda function

In [183]:
import heapq

class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:

        def euc_dist(x, y):
            return x**2 + y**2
        
        minHeap = []
        for x, y in points:
            minHeap.append([euc_dist(x, y), x, y])
        heapq.heapify(minHeap)
        res = []
        while k > 0:
            dist, x, y = heapq.heappop(minHeap)
            res.append([x, y])
            k-=1
        return res

    def kClosest(self, points, K):
        points.sort(key = lambda P: P[0]**2 + P[1]**2)
        return points[:K]
    
points = [[2,2],[2,2],[3,3],[2,-2],[1,1]]
k = 4
s = Solution()
s.kClosest(points, k)

[[1, 1], [2, -2], [2, 2], [2, 2]]

## (3)  Longest Substring Without Repeating Characters

Given a string s, find the length of the longest 
substring
 without repeating characters.

 

Example 1:
```
Input: s = "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.
```
Example 2:
```
Input: s = "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.
```
Example 3:
```
Input: s = "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.
Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.
```

Constraints:

* 0 <= s.length <= 5 * 104
* s consists of English letters, digits, symbols and spaces.

Big O

* Time $O(n)$
* space $O(n)$

**key takeaway**

use **Sliding Window** method, 2 pointers left and right, if right pointer find duplicates, shift left pointer until it not longer contain duplicates

In [132]:
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charSet = set()
        l = 0
        res = 0
        for r in range(len(s)): # iterate right pointer
            # if it is repeat move left forward
            while s[r] in charSet:
                charSet.remove(s[l])
                l+=1
            charSet.add(s[r])
            res = max(res, r - l + 1)
        return res

s = Solution()
#s.lengthOfLongestSubstring("abcabcbb")
#s.lengthOfLongestSubstring("bbbbb")
#s.lengthOfLongestSubstring("pwwkew")
s.lengthOfLongestSubstring("dvdf")

3

## (15) 3Sum

Given an integer array nums, return all the triplets `[nums[i], nums[j], nums[k]]` such that `i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.`

Notice that the solution set must not contain duplicate triplets.
 

Example 1:
```
Input: nums = [-1,0,1,2,-1,-4]
Output: [[-1,-1,2],[-1,0,1]]
Explanation: 
nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
The distinct triplets are [-1,0,1] and [-1,-1,2].
Notice that the order of the output and the order of the triplets does not matter.
```
Example 2:
```
Input: nums = [0,1,1]
Output: []
Explanation: The only possible triplet does not sum up to 0.
```
Example 3:
```
Input: nums = [0,0,0]
Output: [[0,0,0]]
Explanation: The only possible triplet sums up to 0.
``` 

Constraints:

* 3 <= nums.length <= 3000
* -105 <= nums[i] <= 105

**Intuit**
* The extension of Two-sum and Two sum II problem
* In Two sum, we leverage hashtable
* In Two sum II, we 

Big O:
* Two pointers
    * Time $O(nlogn) + O(n^2)$ ==> $O(n^2)$
    * Space $O(n)$ depending on the sorting algorithm
* Hashset

**key takeaway**
* Use 2 pointers (similar to two sum II)
    * Sort number list
    * set number as the first element, try to solve two sum II on the remaining list
    * Iter first number, solve two sum until find the 3 index
* Use hashset

In [38]:
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 3:
            if sum(nums) == 0:
                return [nums]
            else:
                return []
        
        res = []
        sorted_nums = sorted(nums)
        for i in range(len(sorted_nums)-2):
            num = sorted_nums[i]
            # skip duplicated number
            if i > 0 and num == sorted_nums[i-1]:
                continue

            l = i+1
            r = len(sorted_nums)-1
            while l < r:
                threesum = num + sorted_nums[l] + sorted_nums[r]
                if threesum > 0:
                    r-=1
                elif threesum <0:
                    l+=1
                else:
                    res.append([num, sorted_nums[l], sorted_nums[r]])
                    l+=1
                    # avoid duplicates
                    while nums[l] == nums[1 - 1] and l < r:
                        l+=1
        return res

    # Using hashset
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for i in range(len(nums)):
            if nums[i] > 0:
                break
            if i == 0 or nums[i - 1] != nums[i]:
                self.twoSum(nums, i, res)
        return res

    def twoSum(self, nums: List[int], i: int, res: List[List[int]]):
        seen = set()
        j = i + 1
        while j < len(nums):
            complement = -nums[i] - nums[j]
            if complement in seen:
                res.append([nums[i], nums[j], complement])
                while j + 1 < len(nums) and nums[j] == nums[j + 1]:
                    j += 1
            seen.add(nums[j])
            j += 1

nums = [-2,0,1,1,2]
s = Solution()
s.threeSum(nums)

[[-2, 1, 1], [-2, 2, 0]]

## (102) Binary Tree Level Order Traversal

Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).

Example 1:
```
Input: root = [3,9,20,null,null,15,7]
Output: [[3],[9,20],[15,7]]
```
Example 2:
```
Input: root = [1]
Output: [[1]]
```
Example 3:
```
Input: root = []
Output: []
```

Constraints:

* The number of nodes in the tree is in the range [0, 2000].
* -1000 <= Node.val <= 1000

Big O:
* Time $O(N)$
* Space $O(N)$

**key takeaway**
* Using Queue as we always do use BFS, but add a loop and a level counter to get node value for each level, the number of node of the next level is collected at the inner loop

In [238]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        levels = []
        if root is None:
            return levels
        
        level=0
        current_node = root
        queue = deque([current_node])

        while queue:
            levels.append([])
            level_length = len(queue)
            
            for i in range(level_length):
                current_node = queue.popleft()
                levels[level].append(current_node.val)

                if current_node.left is not None:
                    queue.append(current_node.left)
                if current_node.right is not None:
                    queue.append(current_node.right)
            level += 1
        return levels

    # Recusive solution
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        levels = []
        if not root:
            return levels
        
        def helper(node, level):
            # start the current level
            if len(levels) == level:
                levels.append([])

            # append the current node value
            levels[level].append(node.val)

            # process child nodes for the next level
            if node.left:
                helper(node.left, level + 1)
            if node.right:
                helper(node.right, level + 1)
            
        helper(root, 0)
        return levels

## (133) Clone Graph

Given a reference of a node in a connected undirected graph.

Return a deep copy (clone) of the graph.

Each node in the graph contains a value (int) and a list (List[Node]) of its neighbors.
```
class Node {
    public int val;
    public List<Node> neighbors;
}
```

Test case format:

For simplicity, each node's value is the same as the node's index (1-indexed). For example, the first node with val == 1, the second node with val == 2, and so on. The graph is represented in the test case using an adjacency list.

An adjacency list is a collection of unordered lists used to represent a finite graph. Each list describes the set of neighbors of a node in the graph.

The given node will always be the first node with val = 1. You must return the copy of the given node as a reference to the cloned graph.

 

Example 1:

```
Input: adjList = [[2,4],[1,3],[2,4],[1,3]]
Output: [[2,4],[1,3],[2,4],[1,3]]
Explanation: There are 4 nodes in the graph.
1st node (val = 1)'s neighbors are 2nd node (val = 2) and 4th node (val = 4).
2nd node (val = 2)'s neighbors are 1st node (val = 1) and 3rd node (val = 3).
3rd node (val = 3)'s neighbors are 2nd node (val = 2) and 4th node (val = 4).
4th node (val = 4)'s neighbors are 1st node (val = 1) and 3rd node (val = 3).
```
Example 2:

```
Input: adjList = [[]]
Output: [[]]
Explanation: Note that the input contains one empty list. The graph consists of only one node with val = 1 and it does not have any neighbors.
```
Example 3:
```
Input: adjList = []
Output: []
Explanation: This an empty graph, it does not have any nodes.
```
 
Constraints:

* The number of nodes in the graph is in the range [0, 100].
* 1 <= Node.val <= 100
* Node.val is unique for each node.
* There are no repeated edges and no self-loops in the graph.
* The Graph is connected and all nodes can be visited starting from the given node.

**key takeaway**
* Need to use a hashmap to keep track of visited to avoid circular graph
* DFS
* BFS

In [ ]:
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

class Solution:
    def __init__(self):
        self.visited = {}
        
    def cloneGraph_dfs(self, node: 'Node') -> 'Node':
        if not node:
            return node
        
        # If the node was already visited before.
        # Return the clone from the visited dictionary.
        if node in self.visited:
            return self.visited[node]
        
        # else We Create a clone for the given node. and add the node to visited
        # Note that we don't have cloned neighbors as of now, hence [].
        clone = Node(node.val, [])
        
        # The key is original node and value being the clone node.
        self.visited[node] = clone
        
        # Iterate through the neighbors to generate their clones
        # and prepare a list of cloned neighbors to be added to the cloned node.
        if node.neighbors:
            for neightbor in node.neightbors:
                clone.neightbors.append(self.cloneGraph_dfs(neightbor))
                
        return clone
    
    def cloneGraph_bfs(self, node):
        if not node:
            return node
        
        # Dictionary to save the visited node and it's respective clone
        # as key and value respectively. This helps to avoid cycles.
        visited = {}
        
        # Create a queue for BFS and store the initial node
        queue = deque([node])
        visited[node] = Node(node.val, [])
        
        # Start BFS traversal
        while queue:
            current_node = queue.popleft()
            
            # Iterate through all the neighbors of the node
            for neighbor in current_node.neighbors:
                # Clone the neighbor and put in the visited, if not present already
                if neighbor not in visited:
                    visited[neighbor] = Node(neighbor.val, [])
                    queue.append(neighbor)
                    
                # Add the clone of the neighbor to the neighbors of the clone node "n".
                visited[current_node].neighbors.append(visited[neighbor])
        
        # Return the clone of the node from visited.
        return visited[node]

## 150. Evaluate Reverse Polish Notation

You are given an array of strings tokens that represents an arithmetic expression in a Reverse Polish Notation.

Evaluate the expression. Return an integer that represents the value of the expression.

Note that:

The valid operators are '+', '-', '*', and '/'.
Each operand may be an integer or another expression.
The division between two integers always truncates toward zero.
There will not be any division by zero.
The input represents a valid arithmetic expression in a reverse polish notation.
The answer and all the intermediate calculations can be represented in a 32-bit integer.
 

Example 1:
```
Input: tokens = ["2","1","+","3","*"]
Output: 9
Explanation: ((2 + 1) * 3) = 9
```
Example 2:
```
Input: tokens = ["4","13","5","/","+"]
Output: 6
Explanation: (4 + (13 / 5)) = 6
```
Example 3:
```
Input: tokens = ["10","6","9","3","+","-11","*","/","*","17","+","5","+"]
Output: 22
Explanation: ((10 * (6 / ((9 + 3) * -11))) + 17) + 5
= ((10 * (6 / (12 * -11))) + 17) + 5
= ((10 * (6 / -132)) + 17) + 5
= ((10 * 0) + 17) + 5
= (0 + 17) + 5
= 17 + 5
= 22
``` 

Constraints:

* 1 <= tokens.length <= 104
* tokens[i] is either an operator: "+", "-", "*", or "/", or an integer in the range [-200, 200].

Big O:
* Time $O(n)$
* Space $O(n)$

**key takeaway**
* Using a stack data structure to push non operands to the stack, once see an operator, calculate, push the results to the stack, then continue
* Need to first understand Reverse Polish Notation by ask details during the interview
* pay attention to the order of calculation when doing stack pop
* pay attention to the requirement division

In [261]:
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:

        stack = []
        ops = ['+', '-', '*', '/']
        
        def perform_ops(op, a, b):
            if op == '+':
                return a + b
            elif op == '-':
                return a - b
            elif op == '*':
                return a * b
            else:
                return int(a/b)
        
        for i in range(len(tokens)):
            if tokens[i] not in ops:
                stack.append(int(tokens[i]))
            else:
                B = stack.pop()
                A = stack.pop()
                #print(tokens[i], A , B)
                stack.append(perform_ops(tokens[i], A, B))
        return stack.pop()
        
    def evalRPN(self, tokens: List[str]) -> int:
        operations = {
            "+": lambda a, b: a + b,
            "-": lambda a, b: a - b,
            "/": lambda a, b: int(a / b),
            "*": lambda a, b: a * b
        }

        stack = []
        for token in tokens:
            if token in operations:
                number_2 = stack.pop()
                number_1 = stack.pop()
                operation = operations[token]
                stack.append(operation(number_1, number_2))
            else:
                stack.append(int(token))
        return stack.pop()

                
s = Solution()
#tokens = ["2","1","+","3","*"]
#tokens = ["4","13","5","/","+"]
tokens = ["10","6","9","3","+","-11","*","/","*","17","+","5","+"]
s.evalRPN(tokens)

22

# Week 4

## (207) Course Schedule

There are a total of `numCourses` courses you have to take, labeled from `0` to `numCourses - 1`. You are given an array prerequisites where `prerequisites[i] = [ai, bi]` indicates that you must take course bi first if you want to take course ai.

* For example, the pair `[0, 1]`, indicates that to take course 0 you have to first take course 1.
Return true if you can finish all courses. Otherwise, return false.
 

Example 1:
```
Input: numCourses = 2, prerequisites = [[1,0]]
Output: true
Explanation: There are a total of 2 courses to take. 
To take course 1 you should have finished course 0. So it is possible.
```
Example 2:
```
Input: numCourses = 2, prerequisites = [[1,0],[0,1]]
Output: false
Explanation: There are a total of 2 courses to take. 
To take course 1 you should have finished course 0, and to take course 0 you should also have finished course 1. So it is impossible.
``` 

Constraints:

* 1 <= numCourses <= 2000
* 0 <= prerequisites.length <= 5000
* prerequisites[i].length == 2
* 0 <= ai, bi < numCourses
* All the pairs prerequisites[i] are unique.

Big O:

* Time $O(m + n)$  n is number of course, m is the size of prerequisites
* Space $O(m + n)$ n is number of course, m is the size of prerequisites

**Intuition**

We can see that we have been given certain courses with some dependencies between them. The dependencies are expressed as pairs, which provides some hints for framing the problem in terms of a graph.

If we regard each course as a node and draw an edge from bi to ai for any prerequisite `[ai, bi]`(to indicate that course bi should be completed before taking course ai), we get a directed graph.

If there is a cycle in this directed graph, it suggests that we will not be able to finish all of the courses. Otherwise, we can perform a topological sort of the graph to determine the order in which all of the courses can be finished.

As a result, the problem is reduced to determining whether a cycle occurs in a graph. If there is a cycle, we must return `false`. If not, we return `true`.

**key takeaway**
* Graph problem, draw a graph based on the prerequisites where each prerequisite represent an edge, the goal of the problem is to determine for each node whether we can complete (the course) based on the direction of edge
* Use DFS
    * Generate a adjacency list
    * Created a visited set to detect circular graph
* USE Kahn's algorithm to get topological ordering (using BFS [explain](https://leetcode.com/problems/course-schedule/editorial/)

In [ ]:
class Solution:
    # DFS 
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # map each course to prereq adjacency list
        preMap = {i:[] for i in range(numCourses)}
        for crs, pre in prerequisites:
            preMap[crs].append(pre)
        
        visitSet = Set()
        def dfs(crs):
            if crs in visitSet: # we detected a loop
                return False
            if preMap[crs] == []: # no prerequisite
                return True
            
            visitSet.add(crs)
            for pre in preMap[crs]:
                 # recursively run dfs on prereq if any return false (cannot be completed), return false   
                if not dfs(pre): return False 
            visitSet.remove(crs)
            # after ran this course, we set its adjacency list to empty so it will return True for other path
            preMap[crs] = []
            return True
        
        # we run DFS for all courses to ensure
        for crs in range(numCourses):
            if not dfs(crs): return False
        return True
    
    # Kahn's algorithm
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = [0]*numCoures
        adj = [[] for _ in range(numCoures)]
        
        for crs, pre in prerequisites:
            adj[crs].append(pre)
            indegree[crs] += 1
        
        queue = deque()
        for i in the range(numCourses):
            if indegree[i] == 0:
                queue.append(i)
        
        nodesVisited = 0
        while queue:
            node = queue.popleft()
            nodesVisited += 1
            
            for neightbor in adj(node):
                indegree[neightbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)
        
        return nodesVisited == numCourses

## (208) Implement Trie (Prefix Tree)

A `trie` (pronounced as "try") or prefix tree is a tree data structure used to efficiently store and retrieve keys in a dataset of strings. There are various applications of this data structure, such as autocomplete and spellchecker.

Implement the Trie class:

* `Trie()` Initializes the trie object.
* `void insert(String word)` Inserts the string word into the trie.
* `boolean search(String word)` Returns true if the string word is in the trie (i.e., was inserted before), and false otherwise.
* `boolean startsWith(String prefix)` Returns true if there is a previously inserted string word that has the prefix prefix, and false otherwise.

Example 1:
```
Input
["Trie", "insert", "search", "search", "startsWith", "insert", "search"]
[[], ["apple"], ["apple"], ["app"], ["app"], ["app"], ["app"]]
Output
[null, null, true, false, true, null, true]

Explanation
Trie trie = new Trie();
trie.insert("apple");
trie.search("apple");   // return True
trie.search("app");     // return False
trie.startsWith("app"); // return True
trie.insert("app");
trie.search("app");     // return True
``` 

Constraints:

* 1 <= word.length, prefix.length <= 2000
* word and prefix consist only of lowercase English letters.
* At most 3 * 104 calls in total will be made to insert, search, and startsWith.

In [262]:
class TrieNode:
    def __init__(self):
        self.children = {}
        self.endOfWord = False

class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        cur = self.root
        for c in word:
            if c not in cur.children:
                cur.children[c] = TriNode()
            cur = cur.children[c]
        cur.endOfWord = True

    def search(self, word: str) -> bool:
        cur = self.root
        
        for c in word:
            if c not in cur.children:
                return False
            cur = cur.children[c]
        return cur.endOfWord
    
    def startsWith(self, prefix: str) -> bool:
        cur = self.root
        
        for c in prefix:
            if c not in cur.children:
                return False
            cur = cur.children[c]
        return True

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

## (322) Coin Change

You are given an integer array coins representing coins of different denominations and an integer amount representing a total amount of money.

Return the fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any combination of the coins, return -1.

You may assume that you have an infinite number of each kind of coin.
 

Example 1:
```
Input: coins = [1,2,5], amount = 11
Output: 3
Explanation: 11 = 5 + 5 + 1
```
Example 2:
```
Input: coins = [2], amount = 3
Output: -1
```
Example 3:
```
Input: coins = [1], amount = 0
Output: 0
```

Constraints:

* 1 <= coins.length <= 12
* 1 <= coins[i] <= 231 - 1
* 0 <= amount <= 104

**Intuition**

* First, we observe that greety method does not work, since starting from the latest coin does not always produce the min number of coins
* Patten of dynamic programming
    * $dp[0] = 0$
    * $dp[1] = 1$
    * $dp[n] = min_{i=0, ..., n-1}dp[n - c_i] + 1$ and $n - c_i >= 0$ where c a choice of coin, means we get min coin count if we can reach n - c with addition one coin c to add to n

Big O:

* Time: $O(s * n)$
* space: $O(s)$

**key takeaway**

* Bottom up dynamic programming: we compute all the min counts for amount up to n, so we run iterations to compute $dp[n]$
* Top down dynamic programming: run recursion and cache the intermediate value

In [30]:
class Solution:
    
    # bottom up
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount+1] * (amount + 1) #amount + 1 is just a max value as default
        dp[0] = 0
        
        for a in range(1, amount + 1):
            for c in coins:
                # this condition is not needed since a - c will always has min value of 0 but add for clarify
                if a - c >= 0: 
                    dp[a] = min(dp[a], 1 + dp[a - c])
        return dp[amount] if dp[amount] != amount + 1 else - 1
    
    # top down
    def coinChange(self, coins: List[int], amount: int) -> int:
        
        def dfs(rem, cache):
            if rem < 0:
                return -1
            if rem == 0:
                return 0
            if rem in cache:
                return cache[rem]
            
            min_cost = amount + 1
            for coin in coins:
                res = dfs(rem - coin, cache)
                #if res != -1:
                min_cost = min(min_cost, res+1)
                cache[rem] = min_cost
            print(cache)
            return cache[rem]
        
        ans = dfs(amount, {}) 
        return ans if ans != amount + 1 else -1 

s = Solution()
#coins = [1, 2, 5]
coins = [2]
amount = 3
s.coinChange(coins, amount)

{1: 0}
{1: 0, 3: 1}


1

## (238) Product of Array Except Self

Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation.

 

Example 1:
```
Input: nums = [1,2,3,4]
Output: [24,12,8,6]
```
Example 2:
```
Input: nums = [-1,1,0,-3,3]
Output: [0,0,9,0,0]
``` 

Constraints:

* 2 <= nums.length <= 105
* -30 <= nums[i] <= 30
* The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

Big O:
* Time $O(N)$
* Space $O(N)$

**Key Takeaway**
* Method 1 use 2 array, one store prefix, one store surfix, multiple element by element
* Method 2: to save more space, use 1 array to store prefix, calculate surfix on the fly mutliply the array element

In [34]:
class Solution:
    # using 2 arrays, create prefix and surfix arrays and multiple to get answer array
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # Calculate prefix product array
        prefix = [1]*len(nums)
        for i in range(len(nums)):
            if i == 0:
                continue
            else:
                prefix[i] = prefix[i-1]*nums[i-1]
        
        # Calculate surfix product array
        surfix = [1]*len(nums)
        for i in range(len(nums)-1, -1, -1):
            if i == len(nums)-1:
                continue
            else:
                surfix[i] = surfix[i+1] * nums[i+1]
        return [a*b for a,b in zip(prefix, surfix)]
        
    # to achive space O(1), we create answer array on the fly
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        answer = [1]*len(nums)
        # Calculate prefix product array
        for i in range(len(nums)):
            if i == 0:
                continue
            else:
                answer[i] = answer[i-1]*nums[i-1]
                
        # Calculate surfix and multiple the prefix
        cumPrd = 1
        for i in reversed(range(len(nums))):
            if i == len(nums) - 1:
                continue
            else:
                cumPrd *= nums[i+1] # cum prod to the right side
                answer[i] = answer[i] * cumPrd
        return answer
                          
        
    
s = Solution()
nums = [1,2,3,4]
s.productExceptSelf(nums)

[24, 12, 8, 6]

## (155) Min Stack

Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

Implement the MinStack class:

* MinStack() initializes the stack object.
* void push(int val) pushes the element val onto the stack.
* void pop() removes the element on the top of the stack.
* int top() gets the top element of the stack.
* int getMin() retrieves the minimum element in the stack.
You must implement a solution with O(1) time complexity for each function.

Example 1:
```
Input
["MinStack","push","push","push","getMin","pop","top","getMin"]
[[],[-2],[0],[-3],[],[],[],[]]

Output
[null,null,null,null,-3,null,0,-2]

Explanation
MinStack minStack = new MinStack();
minStack.push(-2);
minStack.push(0);
minStack.push(-3);
minStack.getMin(); // return -3
minStack.pop();
minStack.top();    // return 0
minStack.getMin(); // return -2
```
 

Constraints:

* -231 <= val <= 231 - 1
* Methods pop, top and getMin operations will always be called on non-empty stacks.
* At most 3 * 104 calls will be made to push, pop, top, and getMin.

Big O
* Time O(1) is the requirement
* Space O(N)

**Key takeaway**
* Using a another stack to store the min value of stack at the position

In [37]:
class MinStack:

    def __init__(self):
        self.stack = []
        self.min = []
        

    def push(self, val: int) -> None:
        self.stack.append(val) 
        if len(self.min) == 0:
            self.min.append(val)
        else:
            self.min.append(min(self.min[-1], val))

    def pop(self) -> None:
        self.stack.pop()
        self.min.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min[-1]

obj = MinStack()
obj.push(-2)
obj.push(0)
obj.push(-3)
param_3 = obj.top()
param_4 = obj.getMin()
print(param_3, param_4)

-3 -3


In [40]:
[1, 2, 3].pop()

3

## (98) Validate Binary Search Tree

Given the root of a binary tree, determine if it is a valid binary search tree (BST).

A valid BST is defined as follows:

* The left  subtree of a node contains only nodes with keys less than the node's key.
* The right subtree of a node contains only nodes with keys greater than the node's key.
* Both the left and right subtrees must also be binary search trees.
 

Example 1:

```
Input: root = [2,1,3]
Output: true
```
Example 2:

```
Input: root = [5,1,4,null,null,3,6]
Output: false
Explanation: The root node's value is 5 but its right child's value is 4.
```

Constraints:

* The number of nodes in the tree is in the range [1, 104].
* -231 <= Node.val <= 231 - 1

**Key takeaway**

* Resurcive: set lower and upper bound based on left or right (binary search), recursive validate subtree
* Iterative: Using a stack, set lower and upper bound based on left or right (binary search), iterative validate

In [44]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    
    # Recursive
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        def validate(node, lower=-math.inf, upper=math.inf):
            if node is none:
                return True
            if node.val <= lower or node.val >= upper:
                return False
            return validate(node.left, lower, node.val) and validate(node.right, node.val, upper)
        return validate(root)    
   
    # Iterative
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        
        stack = [(root, -math.inf, math.inf)]
        while stack:
            root, lower, upper = stack.pop()
            if not root:
                contine
            val = root.val
            if val <=lower or val >= upper:
                return False
            stack.append((root.right, val, upper))
            stack.append((root.left, lower, val))
        return True

## (200) Number of Islands

Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.

An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

 

Example 1:
```
Input: grid = [
  ["1","1","1","1","0"],
  ["1","1","0","1","0"],
  ["1","1","0","0","0"],
  ["0","0","0","0","0"]
]
Output: 1
```
Example 2:
```
Input: grid = [
  ["1","1","0","0","0"],
  ["1","1","0","0","0"],
  ["0","0","1","0","0"],
  ["0","0","0","1","1"]
]
Output: 3
``` 

Constraints:

* m == grid.length
* n == grid[i].length
* 1 <= m, n <= 300
* grid[i][j] is '0' or '1'.

In [15]:
from collections import deque

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def valid(row, col):
            return 0 <= row < m and 0 <= col < n
        
        matrix = [row[:] for row in grid]
        m = len(matrix)
        n = len(matrix[0])
        islands = 0
        visited = set()
        directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]

        def bfs(r, c):
            queue = deque()
            queue.append((r, c))
            visited.add((r, c))
            
            while queue:
                r, c = queue.popleft()
                for dr, dc in directions:
                    nr, nc = r+dr, c+dc
                    if valid(nr, nc) and (nr, nc) not in visited and matrix[nr][nc] == '1':
                        visited.add((nr, nc))
                        queue.append((nr, nc))

        
        for row in range(m):
            for col in range(n):
                if matrix[row][col] == '1' and (row, col) not in visited:
                    bfs(row, col)
                    islands += 1

        return islands
                    
    def numIslands(self, grid: List[List[str]]) -> int:
        def valid(row, col):
            return 0 <= row < m and 0 <= col < n
        
        matrix = [row[:] for row in grid]
        m = len(matrix)
        n = len(matrix[0])
        islands = 0
        visited = set()
        directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]

        def dfs(r, c):
            stack = []
            stack.append((r, c))
            visited.add((r, c))
            
            while len(stack)>0:
                r, c = stack.pop()
                for dr, dc in directions:
                    nr, nc = r+dr, c+dc
                    if valid(nr, nc) and (nr, nc) not in visited and matrix[nr][nc] == '1':
                        visited.add((nr, nc))
                        queue.append((nr, nc))

        
        for row in range(m):
            for col in range(n):
                if matrix[row][col] == '1' and (row, col) not in visited:
                    dfs(row, col)
                    islands += 1

        return islands
grid = [
  ["1","1","1","1","0"],
  ["1","1","0","1","0"],
  ["1","1","0","0","0"],
  ["0","0","0","1","1"]
]
s = Solution()
s.numIslands(grid)

2

## (994) Rotting Oranges

You are given an m x n grid where each cell can have one of three values:

* 0 representing an empty cell,
* 1 representing a fresh orange, or
* 2 representing a rotten orange.

Every minute, any fresh orange that is 4-directionally adjacent to a rotten orange becomes rotten.

Return the minimum number of minutes that must elapse until no cell has a fresh orange. If this is impossible, return -1

Example 1:

```
Input: grid = [[2,1,1],[1,1,0],[0,1,1]]
Output: 4
```
Example 2:
```
Input: grid = [[2,1,1],[0,1,1],[1,0,1]]
Output: -1
Explanation: The orange in the bottom left corner (row 2, column 0) is never rotten, because rotting only happens 4-directionally.
```
Example 3:
```
Input: grid = [[0,2]]
Output: 0
Explanation: Since there are already no fresh oranges at minute 0, the answer is just 0.
``` 

Constraints:

* m == grid.length
* n == grid[i].length
* 1 <= m, n <= 10
* grid[i][j] is 0, 1, or 2.

In [38]:
import math

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        
        m = len(grid)
        n = len(grid[0])
        fresh_count = 0
        visited = set()
        direction = [[-1, 0], [0, -1], [1, 0], [0, 1]]
        
        if m == 0:
            return -1
        
        for r in range(m):
            for c in range(n):
                if grid[r][c] == 1:
                    fresh_count += 1
        
        def valid(row, col):
            return 0 <= row < m and 0 <= col < n 
        
        
        def bfs(r, c, fresh_count):
            queue = deque()
            queue.append((r, c))
            mins = 0
            
            while queue:
                r, c = queue.popleft()
                for dr, dc in direction:
                    count = 0
                    nr, nc = r+dr, c+dc
                    if valid(nr, nc):
                        if grid[nr][nc] == 1 and (nr, nc) not in visited: 
                            visited.add((nr, nc))
                            queue.append((nr, nc))
                            count+=1
                            fresh_count-=1
                mins = math.ceil(count/4)
                
            return mins, fresh_count
        
        for row in range(m):
            for col in range(n):
                if grid[row][col] == 2:
                    mins, fresh_count = bfs(row, col, fresh_count)
        print(fresh_count)
        return mins if fresh_count == 0 else -1

s = Solution()
grid = [[2,1,1],[1,1,0],[0,1,1]]
#grid = [[0, 2]]
#grid = [[2,1,1],[0,1,1],[1,0,1]]
s.orangesRotting(grid)

0


0

In [79]:
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        
        # get num of rows
        rows = len(grid)
        
        # if grid is empty return -1
        if rows == 0:
            return -1
        
        # Get num of column 
        cols = len(grid[0])
        
        def valid(row, col, rows, cols):
            return 0 <= row < rows and 0 <= col < cols 
        
        # define direction
        direction = [(-1, 0), (0, -1), (1, 0), (0, 1)]
        
        
        # queue for rotten
        rotten = deque()
        
        # fresh cnt
        fresh_count = 0
        
        # visit cells to get rotten to the queue and get fresh count
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 2:
                    rotten.append((r, c))
                elif grid[r][c] == 1:
                    fresh_count += 1
        
        # minute passed
        minutes_passed = 0
        
        # If there are rotten oranges in the queue and there are still fresh oranges in the grid keep looping
        while rotten and fresh_count > 0:
            
            # update the number of minutes passed
            # it is safe to update the minutes by 1, since we visit oranges level by level in BFS traversal.
            minutes_passed += 1
        
            # process roten orange on the current level
            for _ in range(len(rotten)):
                r, c = rotten.popleft()
                
                # visit all adjacent cells
                for dr, dc in direction:
                    nr, nc = r+dr, c+dc
                    # ignore the cell if it is out of the grid boundary
                    #if nr < 0 or nr == rows or nc < 0 or nc == cols:
                    #    continue
                    # ignore the cell if it is empty '0' or visited before '2'
                    #if grid[nr][nc] == 0 or grid[nr][nc] == 2:
                    #    continue
                        
                    # update the fresh oranges count
                    #fresh_count -= 1
                    
                    # mark the current fresh orange as rotten
                    #grid[nr][nc] = 2
                    
                    # add the current rotten to the queue
                    #rotten.append((nc, nr))
                    
                    #if not valid(nr, nc, rows, cols): 
                    #    continue
                    #if grid[nr][nc] == 0 or grid[nr][nc] == 2:
                    #    continue
                    if valid(nr, nc, rows, cols) and grid[nr][nc] == 1:
                        fresh_count -= 1
                        grid[nr][nc] = 2
                        rotten.append((nr, nc))
            print(minutes_passed, fresh_count)
        
        return minutes_passed if fresh_count == 0 else -1

s = Solution()
grid = [[2,1,1],[1,1,0],[0,1,1]]
s.orangesRotting(grid)

1 4
2 2
3 1
4 0


4

In [71]:
from collections import deque

# Time complexity: O(rows * cols) -> each cell is visited at least once
# Space complexity: O(rows * cols) -> in the worst case if all the oranges are rotten they will be added to the queue

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        
        # number of rows
        rows = len(grid)
        if rows == 0:  # check if grid is empty
            return -1
        
        # number of columns
        cols = len(grid[0])
        
        # keep track of fresh oranges
        fresh_cnt = 0
        
        # queue with rotten oranges (for BFS)
        rotten = deque()
        
        # visit each cell in the grid
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 2:
                    # add the rotten orange coordinates to the queue
                    rotten.append((r, c))
                elif grid[r][c] == 1:
                    # update fresh oranges count
                    fresh_cnt += 1
        
        # keep track of minutes passed.
        minutes_passed = 0
        
        # If there are rotten oranges in the queue and there are still fresh oranges in the grid keep looping
        while rotten and fresh_cnt > 0:

            # update the number of minutes passed
            # it is safe to update the minutes by 1, since we visit oranges level by level in BFS traversal.
            minutes_passed += 1
            
            # process rotten oranges on the current level
            for _ in range(len(rotten)):
                x, y = rotten.popleft()
                
                # visit all the adjacent cells
                for dx, dy in [(1,0), (-1,0), (0,1), (0,-1)]:
                    # calculate the coordinates of the adjacent cell
                    xx, yy = x + dx, y + dy
                    # ignore the cell if it is out of the grid boundary
                    if xx < 0 or xx == rows or yy < 0 or yy == cols:
                        continue
                    # ignore the cell if it is empty '0' or visited before '2'
                    if grid[xx][yy] == 0 or grid[xx][yy] == 2:
                        continue
                        
                    # update the fresh oranges count
                    fresh_cnt -= 1
                    
                    # mark the current fresh orange as rotten
                    grid[xx][yy] = 2
                    
                    # add the current rotten to the queue
                    rotten.append((xx, yy))
            print(minutes_passed, fresh_cnt)


        
        # return the number of minutes taken to make all the fresh oranges to be rotten
        # return -1 if there are fresh oranges left in the grid (there were no adjacent rotten oranges to make them rotten)
        return minutes_passed if fresh_cnt == 0 else -1

s = Solution()
grid = [[2,1,1],[1,1,0],[0,1,1]]
s.orangesRotting(grid)

1 4
2 2
3 1
4 0


4